### PR TITLE
Allow to build on apple silicon

### PR DESCRIPTION
**Pull request by MrZammler**
##### Summary

This PR makes some small adjustments to `netdata-installer.sh` to allow to build on Macs with apple silicon (arm). It basically provides the paths to libraries and includes from homebrew.

##### Component Name

Installer

##### Test Plan

Need at least the following packages from homebrew installed:

```
autoconf
automake
bash
cmake
json-c
libtool
libuv
lz4
m4
openssl@1.1
ossp-uuid
pkg-config
```

After that, try running `netdata-installer.sh` specifying a local directory for install, e.g. `/Users/username/install_netdata`. It should build netdata, e.g.:

```
bash-5.1$ uname -a
Darwin MacBook-Air.local 20.5.0 Darwin Kernel Version 20.5.0: Sat May  8 05:10:31 PDT 2021; root:xnu-7195.121.3~9/RELEASE_ARM64_T8101 arm64
bash-5.1$ ../install_netdata/netdata/usr/sbin/netdata -W buildinfo
Version: netdata v1.31.0-420-ga97f06a6e
Configure options:  '--prefix=/Users/mrzammler/stuff/install_netdata/netdata/usr' '--sysconfdir=/Users/mrzammler/stuff/install_netdata/netdata/etc' '--localstatedir=/Users/mrzammler/stuff/install_netdata/netdata/var' '--libexecdir=/Users/mrzammler/stuff/install_netdata/netdata/usr/libexec' '--libdir=/Users/mrzammler/stuff/install_netdata/netdata/usr/lib' '--with-zlib' '--with-math' '--with-user=netdata' '--disable-lto' '--with-bundled-lws' '--with-bundled-protobuf' '--with-bundled-libJudy' 'CFLAGS= -I/opt/homebrew/include/ -I/opt/homebrew/opt/openssl/include/' 'LDFLAGS= -L/opt/homebrew/lib/ -L/opt/homebrew/opt/openssl/lib/'
Features:
    dbengine:                   YES
    Native HTTPS:               YES
    Netdata Cloud:              YES 
    ACLK Next Generation:       YES
    ACLK-NG New Cloud Protocol: YES
    ACLK Legacy:                YES
    TLS Host Verification:      YES
Libraries:
    jemalloc:                NO
    JSON-C:                  YES
    libcap:                  NO
    libcrypto:               YES
    libm:                    YES
    LWS:                     YES static v3.2.2
    mosquitto:               YES
    tcalloc:                 NO
    zlib:                    YES
Plugins:
    apps:                    NO
    cgroup Network Tracking: NO
    CUPS:                    YES
    EBPF:                    NO
    IPMI:                    NO
    NFACCT:                  NO
    perf:                    NO
    slabinfo:                NO
    Xen:                     NO
    Xen VBD Error Tracking:  NO
Exporters:
    AWS Kinesis:             NO
    GCP PubSub:              NO
    MongoDB:                 NO
    Prometheus Remote Write: NO
```

##### Additional Information

* The installer should also already handle intel mac's, if anyone has some time and one, we should test.
* There are various things that can/should be adjusted for macs, to make it more proper for release, e.g:

```
[/Users/mrzammler/stuff/netdata]$ tar xf /var/folders/5s/s8xbf1s93xz5k611dr55wmpr0000gp/T/netdata-go-XXXXXX.vpiHv7Et/go.d.plugin-v0.30.0.darwin-arm64.tar.gz 
tar: Error opening archive: Unrecognized archive format
 FAILED   
```

* We should check various installer options.
* If needed, kickstart could be adjusted to install brew dependencies.
* There are many warnings during compile.

### CODE REVIEW

File: netdata-installer.sh [hunk 1]
```diff
@@ -26,10 +26,15 @@ if [ "${NETDATA_SOURCE_DIR}" != "${INSTALLER_DIR}" ] && [ "${INSTALLER_DIR}" !=
...
-if [ "$(uname -s)" = 'Darwin' ] && [ -d /usr/local/opt/openssl/include ]; then
-  export C_INCLUDE_PATH="/usr/local/opt/openssl/include"
-  export LDFLAGS="-L/usr/local/opt/openssl@1.1/lib"
+# Pull in OpenSSL and other includes and libs properly if on macOS
+if [ "$(uname -s)" = 'Darwin' ]; then
+  if [ "$(uname -m)" = "x86_64" ] && [ -d /usr/local/opt/openssl/include ]; then
+    export C_INCLUDE_PATH="/usr/local/opt/openssl/include"
+    export LDFLAGS="-L/usr/local/opt/openssl/lib"
+  elif [ "$(uname -m)" = "arm64" ] && [ -d /opt/homebrew/opt/openssl/include ]; then
+    export CFLAGS="${CFLAGS} -I/opt/homebrew/include/ -I/opt/homebrew/opt/openssl/include/"
```
[the remark this answers is not in the document]
Comment (iigorkarpov) [reply netdata-installer.sh:35]: From what I've read about HomeBrew on Apple Silicon, the situation is even more complicated. The target installation directory is either /opt/homebrew (in case of ARM package/install) or /usr/local/homebrew. It depends on the OS version, Homebrew version, particular package (presumably) and if you use Rosetta or not.
Unfortunately I have no access to M1 mac, so I can't say for sure.

[the remark this answers is not in the document]
Comment (MrZammler) [reply netdata-installer.sh:35]: Ok, will check for a way to find the path. We also need to check that installing in `/usr/local/` as we have in our docs is ok for various versions...

[the remark this answers is not in the document]
Comment (Ferroin) [reply netdata-installer.sh:35]: @MrZammler The docs are getting overhauled to not include the `--install` option alongside releaseing the new kickstart script since it doesn’t need it. I’m happy to discuss alternatives to `/usr/local/netdata` for our default install prefix though.

[the remark this answers is not in the document]
Comment (MrZammler) [reply netdata-installer.sh:35]: So, I tried installing in /usr/local/netdata (requires to run kickstart with sudo, and works). It launched after install and seems to work fine.

But: I got a new netdata user, complete with it's own login. Not sure if this is somethings users would like... I don't know if there is a way to create a user without real login.

And I can't seem to be able to stop it... tried a few combinations for sudo launchctl stop/unload/ etc... and netdatacli.

At this point, if I'm just a casual user checking things out, I would rather not install in /usr/local, without some indication on how to stop/start it, or uninstall.

[the remark this answers is not in the document]
Comment (Ferroin) [reply netdata-installer.sh:35]: stop/start and the user thing should not have anything to do with `/usr/local`. It sounds like Apple has changed things in some way which is breaking our previously (correctly working) handling of a system user and setup for launchd.

**Review by iigorkarpov — DISMISSED** [review]

**Comment by underhood** [comment]
https://learn.netdata.cloud/docs/agent/packaging/installer/methods/macos

following this i get libuv missing error from configure on M1

**Comment by MrZammler** [comment]
> https://learn.netdata.cloud/docs/agent/packaging/installer/methods/macos
> 
> following this i get libuv missing error from configure on M1

Ok, it seems in your setup, a symlink is missing.... I tried again here, uninstalled and re-installed openssl through brew. Apart from a version upgrade (1.1 -> 3), the symlink seems to be created again:

```
mrzammler@MacBook-Air ~ % ls -l /opt/homebrew/opt/*ssl*
lrwxr-xr-x  1 mrzammler  admin  27 Nov 19 09:01 /opt/homebrew/opt/openssl -> ../Cellar/openssl@3/3.0.0_1
lrwxr-xr-x  1 mrzammler  admin  27 Nov 19 09:01 /opt/homebrew/opt/openssl@3 -> ../Cellar/openssl@3/3.0.0_1
```

Our installer checks for `/opt/homebrew/opt/openssl` and if found it will setup extra `CFLAGS` and `LDFLAGS`, that will also take care of `libuv` includes and libs. I'm not sure what's causing the symlink not to appear in your case, we need to search more...

Also fun fact: If running a x86 terminal (I'm using a third party one) that runs under rosseta, `uname -m` will return `x86_64`...

**Comment by underhood** [comment]
@MrZammler I will try to reset the system to clean out of factory state and start again to see if symlink is created or not.

**Comment by underhood** [comment]
So i started with out of the box system m1 Mac:
Logged in first time:
`/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`

Homebrew postinstall steps asked for by the installer:
```bash
echo 'eval "$(/opt/homebrew/bin/brew shellenv)"' >> /Users/m1/.zprofile
eval "$(/opt/homebrew/bin/brew shellenv)"
```
Install dependencies as netdata docs say here https://learn.netdata.cloud/docs/agent/packaging/installer/methods/macos
```
brew install ossp-uuid autoconf automake pkg-config libuv lz4 json-c openssl@1.1 libtool cmake
```

download source:
```
git clone https://github.com/mrzammler/netdata.git --recursive 
cd netdata
git checkout -b apple-silicon --track
```
Run our installer:
`./netdata-installer.sh --install /Users/m1/aaaa `


build fails with:
```
....
checking for resolv.h... yes
checking for accept4... no
checking operating system... macos with id 3
checking if compiler needs -Werror to reject unknown flags... yes
checking whether pthreads work with -pthread... yes
checking for joinable pthread attribute... PTHREAD_CREATE_JOINABLE
checking if more special flags are required for pthreads... -D_THREAD_SAFE
checking for PTHREAD_PRIO_INHERIT... yes
checking for pthread_getname_np in -lpthread... yes
checking for sin in -lm... yes
checking if libm should be used... yes
checking for uv_fs_scandir_next in -luv... no
configure: error: libuv required but not found. Try installing 'libuv1-dev' or 'libuv-devel'.
 FAILED   
```

**Comment by iigorkarpov** [comment]
@underhood I ran into the same issue on Intel Mac with Macports. I had to add LDFLAGS pointing to libuv, as far as I remember.

**Comment by underhood** [comment]
@iigorkarpov yes we can fix it manually by providing paths of course. But point we try to achieve is for build to work out of box on Mac and if that is not possible -> updating the netdata documentation so that such manual steps are documented.

**Comment by MrZammler** [comment]
Ahhh, perhaps.... @underhood can you check if you install `openssl` from brew, instead of `openssl@1.1` ?

**Comment by underhood** [comment]
i did `brew install openssl@1.1` and then to try it out i did also `brew install openssl` which updated it to 3 but same error from installer

**Comment by MrZammler** [comment]
Hi!, an update:

Ok, from Timo's testing (thanks!) it seems that now there are 2 packages in brew for openssl: 

openssl@1.1 (https://formulae.brew.sh/formula/openssl@1.1#default), and openssl@3 (https://formulae.brew.sh/formula/openssl@3#default). openssl@3 is also linked to a package named openssl (Also known as: openssl on the webpage above.)

If you run `brew install openssl` or `brew install openssl@3`, it will install openssl@3 and will create a symlink to it (`/opt/homebrew/opt/openssl -> ../Cellar/openssl@3/3.0.0_1`). Installing openssl@1.1 will not. This must have changed at some point after I did initial tests on my mac, where openssl@1.1 would install also a symlink from openssl.

Now, with this PR, we check the existence of `/opt/homebrew/opt/openssl/include/` to determine if we should add includes and libs for silicon macs.

Two choices then:

1) We install via `install-required-packages.sh` openssl rather than openssl@1.1. The PR's current status is to do that.
2) If for some reason openssl@3 does not work with Netdata (we need to check...), we should install openssl@1.1, and check for the existence of `/opt/homebrew/opt/openssl@1.1/include/` and use that.

So, this last commit (https://github.com/netdata/netdata/pull/11707/commits/28743b802d6864b2eb21647b021c27d982b911b7) makes 3 changes:

* Install `json-c` (`json-c-devel` does not seem to exist in brew).
* Install `openssl` instead of `openssl@1.1`.
* Runs when in macos the `install-required-packages.sh` without root. Running it with sudo results in error: `This script does not support running as EUID 0 on macOS. Please run it as a regular user.`

How to test (and we need to test in both x86 and arm) on a completely empty mac (no brew):

* Install brew.
* Install bash via brew.
* Clone this repository, switch to branch `apple-silicon`.
* Update the git submodules: (`git submodule update --init --recursive`)
* `cd packaging/installer`
* Run: `NETDATA_LOCAL_TARBALL_OVERRIDE_DEPS_SCRIPT=/Users/m1/netdata/packaging/installer/install-required-packages.sh ./kickstart.sh --dont-start-it --install /Users/m1/install_netdata`. Adjust for the path to the cloned repository, and also it would be good to install to a local directory. Installing to e.g. `/usr/local` requires some more searching.
* The above should take care of installing the required packages through brew. It should not error out.
* But when it comes to running `netdata-install.sh` it will do so from whatever tarball it has downloaded, and it will fail. We need to run `netdata-install.sh` from this repository, so:
* `cd ../../`
* Run: `sudo ./netdata-installer.sh --auto-update --dont-start-it --install /Users/m1/install_netdata`. This should compile netdata, and install it.

But still, there are issues we need to look at before considering installing netdata with kickstart for macos to be user-friendly:

* Check if we can install in `/usr/local/` and see if `launchd` will play nicely. We could consider e.g. `/opt/` instead.
* Check if the service can be at least started/stopped.
* Check if we need an extra `netdata` user on Mac.

**Review by underhood — APPROVED** [review]

**Review by DShreve2 — APPROVED** [review]

File: netdata-installer.sh [hunk 2]
```diff
@@ -26,10 +26,26 @@ if [ "${NETDATA_SOURCE_DIR}" != "${INSTALLER_DIR}" ] && [ "${INSTALLER_DIR}" !=
 fi
 
 # -----------------------------------------------------------------------------
-# Pull in OpenSSL properly if on macOS
-if [ "$(uname -s)" = 'Darwin' ] && [ -d /usr/local/opt/openssl/include ]; then
-  export C_INCLUDE_PATH="/usr/local/opt/openssl/include"
-  export LDFLAGS="-L/usr/local/opt/openssl@1.1/lib"
+# Pull in OpenSSL and other includes and libs properly if on macOS
+if [ "$(uname -s)" = 'Darwin' ]; then
```
Comment (vkalintiris) [inline netdata-installer.sh:30]: @MrZammler FWIW, I think it would be better to use `HOMEBREW_PREFIX` to resolve these paths. If the variable is missing we could try to get it by using `brew`'s `shellenv`. Having said that I think the PR looks good.

**Comment by MrZammler** [comment]
It seems this has gone too long, and conflicts are tricky, or perhaps not safe maybe to solve here, and things in installer and required packages have changed over this period.

Let's continue in https://github.com/netdata/netdata/pull/12048 and/or https://github.com/netdata/netdata/pull/12051

I can do another PR on top of those for arm, or we can put the changes from this PR on those, I'm ok with both.
